### PR TITLE
Initial topography plugin prm polygon

### PR DESCRIPTION
**Pull request by MFraters**
This adds an initial topography plugin which defines the initial topography by defining 2d polygons on the surface and setting it to the desired elevation. After this I will update the ellipsoidal chunk and make a test out of this. Or should I do this within this pull request?

### CODE REVIEW

**Comment by MFraters** [comment]
I added a function to utilities which computes if a point falls within a 2d polygon. I added it there because I noticed that it is a function I use a lot in my plugins, especially for initial conditions. So I thought it might be useful for other people as well.

File: include/aspect/geometry_model/initial_topography_model/prm_polygon.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      private:
+        /**
+         * The values of the topography are stored in a vector.
+         */
+        std::vector<double> topography_values;
+
+        /**
+         * The polygons and their points are stored in this vector.
+         */
+        std::vector<std::vector<std::vector<double> > > point_lists;
```
Comment (bangerth) [inline include/aspect/geometry_model/initial_topography_model/prm_polygon.h]: Since you know the size of the inner vector, can you change the inner vector to `Point<dim>` or `Point<dim-1>` or whatever is the appropriate type?

File: include/aspect/utilities.h [hunk 2]
```diff
@@ -123,6 +123,14 @@ namespace aspect
     }
 
     /**
+     * Given a 2d point and a list of points which form a polygon, computes if the point
+     * falls within the polygon.
+     */
+    template <int dim>
+    bool
+    in_2d_polygon(dealii::Point<2> &point,const std::vector<std::vector<double> > &pointList);
```
Comment (bangerth) [inline include/aspect/utilities.h]: same here -- change it to `vector<Point<2>>`. Make the `point` argument `const`.

For unclear, historical reasons, we seem to usually list the bigger arguments first, so `points_list` (not name) first and `point` second.

Comment (bangerth) [inline include/aspect/utilities.h]: As a follow-up, reversing the arguments then also suggests naming the function `polygon_contains_point`. No need for the "2d" since a polygon is always 2d (otherwise it'd be a polyhedron).

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    value (const Point<dim-1> &p) const
+    {
+      double value = 0;
+      for (unsigned int i = 0; i < point_lists.size(); i++)
+        {
+          Point<2> p1(p[0],p[1]);
+
+          if (aspect::Utilities::in_2d_polygon<dim>(p1, point_lists[i]))
+            {
+              value = topography_values[i];
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: You are continuing the loop over the polygons here. This only makes sense if you expect the point to be in multiple polygons, in which case you return the topography of the last one. I think that's probably unintended -- just say
```
  return topography_values[i];
```
which takes the value from the first polygon. At the end of the function, you can then just say `return 0;`

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  {
+    template <int dim>
+    double
+    PrmPolygon<dim>::
+    value (const Point<dim-1> &p) const
+    {
+      double value = 0;
+      for (unsigned int i = 0; i < point_lists.size(); i++)
+        {
+          Point<2> p1(p[0],p[1]);
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: make it `const`. Move the variable out of the loop.

You may want to add an assertion that double checks that you really are in 3d, since otherwise the access to `p[1]` will be problematic.

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+      {
+        prm.enter_subsection("Initial topography");
+        {
+          prm.enter_subsection("Prm polygon");
+          {
+            prm.declare_entry("Topography parameters",
+                              "",
+                              Patterns::Anything(),
+                              "Set the topography height, end with a |, and set the areas described "
+                              "by the points, separated by commas and coordinates separated by a ':'. "
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: This sentence is not correct grammar. Can you also first describe what the parameter does, rather than jumping straight into a description of its format?

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+          prm.enter_subsection("Prm polygon");
+          {
+            /**
+             * we need to fill the point lists and topography values. They
+             * are stored in the Topography subsection in the Topography parameter.
+             */
+
+            std::string temptopo = prm.get("Topography parameters");
+            std::vector<std::string> temp_topographies = Utilities::split_string_list(temptopo,';');
+            unsigned int temp_topographies_size = temp_topographies.size();
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: make all of these variables `const`.

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        {
+          prm.enter_subsection("Prm polygon");
+          {
+            prm.declare_entry("Topography parameters",
+                              "",
+                              Patterns::Anything(),
+                              "Set the topography height, end with a |, and set the areas described "
+                              "by the points, separated by commas and coordinates separated by a ':'. "
+                              "Seperate each topography feature by a semicolon. For example for two triangular areas of "
+                              "100 and -100 meters high set: '100|0:0,5:5,0:10;-100|10:10,10:15,20:15'. "
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: `|` is sometimes treated as a special character by `ParameterHandler`. I would prefer if you used `:` to separate the elevation from the point list, commas to separate coordinates, and semicolons to separate points. Then use something else to separate polygons -- maybe `&` would be a good choice.

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+            std::string temptopo = prm.get("Topography parameters");
+            std::vector<std::string> temp_topographies = Utilities::split_string_list(temptopo,';');
+            unsigned int temp_topographies_size = temp_topographies.size();
+
+            topography_values.resize(temp_topographies_size,0);
+            point_lists.resize(temp_topographies_size);
+            for (unsigned int i_topo = 0; i_topo < temp_topographies_size; i_topo++)
+              {
+                std::vector<std::string> temp_topography = Utilities::split_string_list(temp_topographies[i_topo],'|');
+                Assert(temp_topography.size() == 2,ExcMessage ("The given line '" + temp_topographies[i_topo] + "' is not correct. "
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: Can you break the ExcMessage onto the next line to shorten the length of the lines here?

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+                                                               "It consists of " + boost::lexical_cast<std::string>(temp_topography.size())
+                                                               + " parts separated by |, but it should only contain 2 parts: "
+                                                               "the height and the coordinate, separated by a |.'"));
+
+                topography_values[i_topo] = Utilities::string_to_double(temp_topography[0]);
+
+                std::vector<std::string> temp_coordinates = Utilities::split_string_list(temp_topography[1],',');
+                unsigned int temp_coordinate_size = temp_coordinates.size();
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: make these `const`.

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+                point_lists[i_topo].resize(temp_coordinate_size,std::vector<double>(2,0));
+
+                for (unsigned int i_coord = 0; i_coord < temp_coordinate_size; i_coord++)
+                  {
+                    point_lists[i_topo][i_coord] = Utilities::string_to_double(Utilities::split_string_list(temp_coordinates[i_coord],':'));
+                    Assert(temp_topography.size() == 2,ExcMessage ("The given coordinate '" + temp_coordinates[i_coord] + "' is not correct. "
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: same here

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+// explicit instantiations
+namespace aspect
+{
+  namespace InitialTopographyModel
+  {
+    ASPECT_REGISTER_INITIAL_TOPOGRAPHY_MODEL(PrmPolygon,
+                                             "prm polygon",
+                                             "A initial topography model which defines the initial "
+                                             "topography by defining 2d polygons on the surface and "
+                                             "setting it to the desired elevation. ")
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: -> ... and prescribing a constant elevation inside each polygon

File: source/utilities.cc [hunk 12]
```diff
@@ -167,6 +167,57 @@ namespace aspect
     }
 
     template <int dim>
+    bool
+    in_2d_polygon(dealii::Point<2> &point,const std::vector<std::vector<double> > &pointList)
+    {
```
Comment (bangerth) [inline source/utilities.cc]: What happens if you call this function with `dim==3`? Does it just ignore the 3rd component of each point in the list?

Comment (bangerth) [inline source/utilities.cc]: `pointList` -> `point_list`.

**Comment by bangerth** [comment]
Nice addition. I think it would be nice to have a test as well.

File: source/geometry_model/initial_topography_model/prm_polygon.cc (hunk 3 above)
Comment (MFraters) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: That is actually the intended behaviour :)

File: source/geometry_model/initial_topography_model/prm_polygon.cc (hunk 7 above)
Comment (MFraters) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: I changed them. I only replaced the | not with a :, but with a >, because the : character looks to me too much as a coordinate thing.

**Comment by MFraters** [comment]
I hope this addresses all comments :)

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+      else
+        p1 = Point<2>(p[0],p[1]);
+
+      for (unsigned int i = 0; i < point_lists.size(); i++)
+        {
+          if (aspect::Utilities::polygon_contains_point<dim>(point_lists[i],p1))
+            {
+              value = topography_values[i];
+            }
+        }
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: The previous comments on this code have been erased during renaming of the function, but to rehash: I said that this is inefficient because it has to go through all of the point lists, and then takes the value of the *last* polygon that matches. You said that this is the intended behavior.

I think I don't understand your use case where you have multiple polygons, and more than one of them actually has the given point inside. If that's the case, why take the *last* polygon's value, rather than the *first*? Taking the first has the advantage that you can just exit the function at this point, rather than having to search through all following polygons.

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+    PrmPolygon<dim>::
+    value (const Point<dim-1> &p) const
+    {
+      double value = 0;
+      Point<2> p1;
+
+      if (dim == 2)
+        p1 = Point<2>(p[0],0);
+      else
+        p1 = Point<2>(p[0],p[1]);
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: You can write lines 36-41 more efficiently as
```
  const Point<2> p1 = (dim == 2 ? Point<2>(p[0],0) : Point<2>(p[0],p[1]));
```

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+      {
+        prm.enter_subsection("Initial topography");
+        {
+          prm.enter_subsection("Prm polygon");
+          {
+            prm.declare_entry("Topography parameters",
+                              "",
+                              Patterns::Anything(),
+                              "Set the topography height and the polygon which should be set to that height. "
+                              "The formate is : \"The topography height > The point list describing a polygon & "
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: formate -> format

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+                              "The next topography height > the next point list describing a polygon.\" "
+                              "The format for the point list describing the polygon is \"x1,y1;x2,y2\". "
+                              "For example for two triangular areas of 100 and -100 meters high set: "
+                              "'100 > 0,0;5,5;0,10 & -100 > 10,10;10,15;20,15'. "
+                              "Units of the height are always in meters. The units of the coorindates are "
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: -> coordinates

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+                              "dependent on the geometry model.");
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: Can you be more specific here? E.g., for the 'box' model, it is interpreted in meters; for spherical models, point coordinates are interpreted as latitude/longitude? Or do I entirely misinterpret what you're trying to say.

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+        {
+          prm.enter_subsection("Prm polygon");
+          {
+            /**
+             * we need to fill the point lists and topography values. They
+             * are stored in the Topography subsection in the Topography parameter.
+             */
+            std::string temptopo = prm.get("Topography parameters");
+            std::vector<std::string> temp_topographies = Utilities::split_string_list(temptopo,'&');
+            unsigned int temp_topographies_size = temp_topographies.size();
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: I think all three of these can be `const`, right?

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+            topography_values.resize(temp_topographies_size,0);
+            point_lists.resize(temp_topographies_size);
+            for (unsigned int i_topo = 0; i_topo < temp_topographies_size; i_topo++)
+              {
+                const std::vector<std::string> temp_topography = Utilities::split_string_list(temp_topographies[i_topo],'>');
+                Assert(temp_topography.size() == 2,ExcMessage ("The given line '" + temp_topographies[i_topo]
+                                                               + "' is not correct. It consists of "
+                                                               + boost::lexical_cast<std::string>(temp_topography.size())
+                                                               + " parts separated by >, but it should only contain "
+                                                               "2 parts: the height and the coordinate, separated "
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: 2 -> two
coordinate -> list of point coordinates

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+                const std::vector<std::string> temp_coordinates = Utilities::split_string_list(temp_topography[1],';');
+                const unsigned int temp_coordinate_size = temp_coordinates.size();
+                point_lists[i_topo].resize(temp_coordinate_size);
+                for (unsigned int i_coord = 0; i_coord < temp_coordinate_size; i_coord++)
+                  {
+                    const std::vector<double> temp_point = Utilities::string_to_double(Utilities::split_string_list(temp_coordinates[i_coord],','));
+                    Assert(temp_point.size() == 2,ExcMessage ("The given coordinate '" + temp_coordinates[i_coord] + "' is not correct. "
+                                                              "It consists of " + boost::lexical_cast<std::string>(temp_topography.size())
+                                                              + " parts separated by :, but it should only contain 2 parts: "
+                                                              "the longitude and latitude of the the coordinate, separated by a ','."));
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: since it's up to other parts of the program to interpret coordinates, maybe say "the two coordinates of the polygon points"?

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+namespace aspect
+{
+  namespace InitialTopographyModel
+  {
+    ASPECT_REGISTER_INITIAL_TOPOGRAPHY_MODEL(PrmPolygon,
+                                             "prm polygon",
+                                             "A initial topography model which defines the initial "
+                                             "topography by defining 2d polygons on the surface and "
+                                             "setting it to the desired elevation and prescribing a "
+                                             "constant elevation inside each polygon. ")
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: This sentence is too long. Maybe:
```
  An initial topography model that defines the initial topography as constant inside each of a set of polygonal 
  parts of the surface. The polygons, and their associated surface elevation, are defined in the
  `Geometry model/Initial topography/Prm polygon' section.
```

File: source/utilities.cc [hunk 22]
```diff
@@ -210,6 +210,61 @@ namespace aspect
     }
 
     template <int dim>
+    bool
+    polygon_contains_point(const std::vector<Point<2> > &point_list, const dealii::Point<2> &point)
```
Comment (bangerth) [inline source/utilities.cc]: break the two arguments onto separate lines

File: include/aspect/utilities.h [hunk 23]
```diff
@@ -132,6 +132,14 @@ namespace aspect
     }
 
     /**
+     * Given a 2d point and a list of points which form a polygon, computes if the point
+     * falls within the polygon.
+     */
+    template <int dim>
+    bool
+    polygon_contains_point(const std::vector<Point<2> > &point_list, const dealii::Point<2> &point);
```
Comment (bangerth) [inline include/aspect/utilities.h]: break the two arguments onto separate lines

File: source/geometry_model/initial_topography_model/prm_polygon.cc (hunk 13 above)
Comment (MFraters) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: You are right it is inefficient. My point was mainly that the last should be chosen. It is more a conceptual thing for me, that you go through the list and 'paint over' in every polygon. I have made it more efficient now, in the sense that I turned around the for loop and put the return inside the if statement. 

Although it currently has my preference that the last one is the final one, but I do not feel very strongly about it. So if you really prefer it to be the other way around I am fine with it.

File: source/geometry_model/initial_topography_model/prm_polygon.cc (hunk 21 above)
Comment (MFraters) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: That is indeed a much better description. Thanks!

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+          {
+            prm.declare_entry("Topography parameters",
+                              "",
+                              Patterns::Anything(),
+                              "Set the topography height and the polygon which should be set to that height. "
+                              "The format is : \"The topography height > The point list describing a polygon & "
+                              "The next topography height > the next point list describing a polygon.\" "
+                              "The format for the point list describing the polygon is \"x1,y1;x2,y2\". "
+                              "For example for two triangular areas of 100 and -100 meters high set: "
+                              "'100 > 0,0;5,5;0,10 & -100 > 10,10;10,15;20,15'. "
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: Thinking about it, this won't render correctly in latex. I think you need to write `$>$` or maybe better, `\textgreater`.

File: include/aspect/geometry_model/initial_topography_model/prm_polygon.h [hunk 25]
```diff
@@ -0,0 +1,79 @@
+/*
+  Copyright (C) 2011, 2012, 2015 by the authors of the ASPECT code.
```
Comment (bangerth) [inline include/aspect/geometry_model/initial_topography_model/prm_polygon.h]: Fix dates: only 2016

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 26]
```diff
@@ -0,0 +1,155 @@
+/*
+  Copyright (C) 2011 - 2016 by the authors of the ASPECT code.
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: fix dates: only 2016

**Comment by bangerth** [comment]
I was going to merge, but then I saw the issue with `>` in latex. Then I thought I'd also point out the issues with the copyright dates. Can you fix these, and squash everything into one commit?

Thinking about it, you probably also want to add an entry to the `changes.h` file, so you get credit where credit is due.

Otherwise ready to merge.

File: source/geometry_model/initial_topography_model/prm_polygon.cc [hunk 27]
```diff
@@ -0,0 +1,155 @@
...
+    value (const Point<dim-1> &p) const
+    {
+      double value = 0;
+      Point<2> p1 = (dim == 2 ?Point<2>(p[0],0) : Point<2>(p[0],p[1]));
+
+      for (unsigned int i = point_lists.size(); i > 0; i--)
+        {
+          if (aspect::Utilities::polygon_contains_point<dim>(point_lists[i-1],p1))
+            {
+              return topography_values[i-1];
```
Comment (bangerth) [inline source/geometry_model/initial_topography_model/prm_polygon.cc]: This order of the loop works for me. You may want to add a comment explaining why this is the order in which you go through the loop this way.

**Comment by MFraters** [comment]
I must say that it currently still is not usable, because no geometry model has implemented the initial topography yet. But that will be the next step :)

About the squashing, I normally squash (fixup actually) all changes in one source and include pair in one commit, and each test in one commit. Is this what you meant, or do you want everything into one commit which contains everything?

Furthermore, I noticed that before adding the last set of changes (after which the tester ran again), the tester failed because three melt tests where taking to long. For now it is fine, but maybe the allowed time per test should be extended a bit, or the tests should be simplified a bit in the future?

**Comment by bangerth** [comment]
Yes, one for code changes, one for changelog, one for tests is totally fine. That's how I often do it as well.

Regarding tester time, that's @tjhei 's call. But I haven't seen this problem before, so I suspect that it doesn't happen very often and in fact only happens if something else is happening in the background.

File: doc/modules/changes.h [hunk 28]
```diff
@@ -6,6 +6,12 @@
  *
  * <ol>
  *
+ * <li> New: There is now an initial topography plugin which reads 
+ * from the prm file polygon definitons and set the initial topography
```
Comment (bangerth) [inline doc/modules/changes.h]: -> definitions

File: doc/modules/changes.h [hunk 29]
```diff
@@ -6,6 +6,12 @@
  *
  * <ol>
  *
+ * <li> New: There is now an initial topography plugin which reads 
+ * from the prm file polygon definitons and set the initial topography 
+ * to be constant withing those polygons.
```
Comment (bangerth) [inline doc/modules/changes.h]: -> within

**Comment by bangerth** [comment]
I was going to merge, because the structure with the 4 commits is also just fine -- in fact, it's a nice way of doing it this way because the function that tests whether a point is inside a polygon is a self-contained addition that can certainly be in its own commit.

But I saw that you have two typos in the changelog. Can you still fix them? (Logically, btw, the second commit with `polygon_contains_point` should come before the first commit because currently the first requires the second -- in other words, you can't put yourself onto the first commit and compile ASPECT -- you will get a compiler error. That's really not an important issue, but it's also really easy to fix with `git rebase -i` by just swapping two lines.)

**Comment by MFraters** [comment]
Thanks! Your are right, that order makes more sense. 

I also noticed that there where still two comments in prm_polygon.h incorrect, so I fixed them (hopefully didn't introduce more spelling mistakes :P )

**Comment by MFraters** [comment]
Please do not merge yet, I found a possible mistake somewhere. I want to check it first.

**Comment by bangerth** [comment]
OK. Let us know when you have it fixed :-)

**Comment by MFraters** [comment]
I found that the name was not of the subsection was not consistent with the interface, and that is now fixed. 

At the same time I was making the implementation of this in the ellipsoidal chunk geometry model, and I found some weird behaviour. But I am now quite sure that doesn't have to do with this part of the code, but with what happens when you use a very course grid with a very large topography. But that is something we might discuss in that pull request.

In short, if the tester is fine with this, I think it is ready to merge :)
